File: src/Other/Internals/ControlToDisplayCodeHostedOnGitHub.cs
using CSHTML5.Internal;
using OpenSilver.Themes.Modern;
using System.Windows;
using System.Windows.Browser;
using System.Windows.Controls;

namespace OpenSilver.Samples.Showcase
{
    public class ControlToDisplayCodeHostedOnGitHub : ContentControl
    {
        string _filePathOnGitHub;
        string _displayedHtmlString;

        public ControlToDisplayCodeHostedOnGitHub()
        {
            Loaded += OnLoaded;

            VerticalContentAlignment = VerticalAlignment.Stretch;
            HorizontalContentAlignment = HorizontalAlignment.Stretch;
        }

        public ControlToDisplayCodeHostedOnGitHub(string absolutePath) : this()
        {
            FilePathOnGitHub = absolutePath;
        }

        string GetHtmlString(string filePath)
        {
            var embedJs = INTERNAL_UriHelper.ConvertToHtml5Path("ms-appx:/Other/embed.js");
            var themeMode = Application.Current.Theme is ModernTheme theme && theme.CurrentPalette == ModernTheme.Palettes.Dark ? "dark" : "github";

            return $"<script src=\"{embedJs}?target={HttpUtility.UrlEncode(filePath)}&style={themeMode}&showBorder=on&showLineNumbers=on&showCopy=on\"></script>";
        }

        void OnLoaded(object sender, RoutedEventArgs e)
        {
            if (!string.IsNullOrEmpty(_filePathOnGitHub))
            {
                string htmlString = GetHtmlString(_filePathOnGitHub);
                DisplayHtmlString(htmlString);
            }
        }

        void DisplayHtmlString(string htmlString)
        {
            Interop.ExecuteJavaScriptVoidAsync($@"
                (function () {{
                    const iframe = document.createElement('iframe');
                    iframe.style=""border:none;width:100%;height:100%;overflow:hidden;display:block;pointer-events:auto"";
                    iframe.onload = function() {{
                        iframe.contentDocument.open();
                        iframe.contentDocument.write('<html><head>{htmlString}</head></html>');
                        iframe.contentDocument.close();
                    }};
                    const currentDiv = document.getElementById(""{(Interop.GetDiv(this) as INTERNAL_HtmlDomElementReference)?.UniqueIdentifier}"");
                    currentDiv.appendChild(iframe);
                    iframe.src = 'about:blank';
                }})();");
            _displayedHtmlString = htmlString;
        }

        public string FilePathOnGitHub
        {
            get => _filePathOnGitHub;
            set
            {
                _filePathOnGitHub = value;

                if (IsLoaded)
                {
                    string htmlString = GetHtmlString(FilePathOnGitHub);
                    if (htmlString != _displayedHtmlString)
                    {
                        DisplayHtmlString(htmlString);
                    }
                }
            }
        }

        public void Refresh()
        {
            if (!string.IsNullOrEmpty(_filePathOnGitHub) && IsLoaded)
            {
                string htmlString = GetHtmlString(_filePathOnGitHub);
                DisplayHtmlString(htmlString);
            }
        }
    }
}

File: src/Other/Internals/ControlToDisplayCodeHostedOnGitHub.fs
namespace OpenSilver.Samples.Showcase

open System.Windows.Browser
open CSHTML5.Internal
open System
open System.Windows
open System.Windows.Controls

type ControlToDisplayCodeHostedOnGitHub() as this =
    inherit ContentControl()

    let mutable _filePathOnGitHub = ""
    let mutable _displayedHtmlString = ""

    do
        this.Loaded.Add(fun args -> this.OnLoaded(this, args))
        this.VerticalContentAlignment <- VerticalAlignment.Stretch
        this.HorizontalContentAlignment <- HorizontalAlignment.Stretch

    public new(absolutePath: string) as this =
        ControlToDisplayCodeHostedOnGitHub()
        then
            this.FilePathOnGitHub <- absolutePath

    member private this.GetHtmlString(filePath : string) =
        let embedJs = INTERNAL_UriHelper.ConvertToHtml5Path("ms-appx:/Other/embed.js")
        let themeMode =
            match Application.Current.Theme with
            | :? OpenSilver.Themes.Modern.ModernTheme as theme when theme.CurrentPalette = OpenSilver.Themes.Modern.ModernTheme.Palettes.Dark ->
                "dark"
            | _ -> "github"
        sprintf
            "<script src=\"%s?target=%s&style=%s&showBorder=on&showLineNumbers=on&showCopy=on\"></script>"
            embedJs (HttpUtility.UrlEncode(filePath)) themeMode

    member private this.OnLoaded(sender : obj, e : RoutedEventArgs) =
        if not (String.IsNullOrEmpty _filePathOnGitHub) then
            let htmlString = this.GetHtmlString _filePathOnGitHub
            this.DisplayHtmlString(htmlString)

    member private this.DisplayHtmlString(htmlString : string) =
        let webView = new WebBrowser()
        webView.NavigateToString(htmlString)
        this.Content <- webView
        _displayedHtmlString <- htmlString

    member this.FilePathOnGitHub
        with get() = _filePathOnGitHub
        and set value =
            _filePathOnGitHub <- value

            if this.IsLoaded then
                let htmlString = this.GetHtmlString this.FilePathOnGitHub
                if htmlString <> _displayedHtmlString then
                    this.DisplayHtmlString(htmlString)

    member this.Refresh() =
        if not (String.IsNullOrEmpty(_filePathOnGitHub)) && this.IsLoaded then
            let htmlString = this.GetHtmlString(_filePathOnGitHub)
            this.DisplayHtmlString(htmlString)
